src/main/java/com/google/jenkins/plugins/credentials/oauth/JsonServiceAccountConfig.java

```diff
@@ -53,36 +63,97 @@
   private static final long serialVersionUID = 6818111194672325387L;
   private static final Logger LOGGER =
       Logger.getLogger(JsonServiceAccountConfig.class.getSimpleName());
-  private String jsonKeyFile;
+  @CheckForNull
+  private String filename;
+  @CheckForNull
+  private SecretBytes secretJsonKey;
+  @Deprecated   // for migration purpose
+  @CheckForNull
+  private transient String prevJsonKeyFile;
   private transient JsonKey jsonKey;
 
   @DataBoundConstructor
-  public JsonServiceAccountConfig(FileItem jsonKeyFile,
-      String prevJsonKeyFile) {
+  public JsonServiceAccountConfig() {}
+
+  /**@param jsonKeyFile uploaded json key file */
+  @DataBoundSetter // Called on form submission, only used when key file is uploaded
+  public void setJsonKeyFile(FileItem jsonKeyFile) {
     if (jsonKeyFile != null && jsonKeyFile.getSize() > 0) {
       try {
         JsonKey jsonKey = JsonKey.load(new JacksonFactory(),
-                jsonKeyFile.getInputStream());
+            jsonKeyFile.getInputStream());
         if (jsonKey.getClientEmail() != null &&
-                jsonKey.getPrivateKey() != null) {
-          try {
-            this.jsonKeyFile = writeJsonKeyToFile(jsonKey);
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "Failed to write json key to file", e);
-          }
+            jsonKey.getPrivateKey() != null) {
+          this.filename = extractFilename(jsonKeyFile.getName());
+          this.secretJsonKey = SecretBytes.fromBytes(jsonKeyFile.get());
         }
       } catch (IOException e) {
         LOGGER.log(Level.SEVERE, "Failed to read json key from file", e);
       }
-    } else if (prevJsonKeyFile != null && !prevJsonKeyFile.isEmpty()) {
-      this.jsonKeyFile = prevJsonKeyFile;
     }
   }
 
-  private String writeJsonKeyToFile(JsonKey jsonKey) throws IOException {
-    File jsonKeyFile = KeyUtils.createKeyFile("key", ".json");
-    KeyUtils.writeKeyToFileEncoded(jsonKey.toPrettyString(), jsonKeyFile);
-    return jsonKeyFile.getAbsolutePath();
+  /** @param filename json key file name.*/
+  @DataBoundSetter
+  public void setFilename(String filename) {
+    String newFilename = extractFilename(filename);
+    if (!Strings.isNullOrEmpty(newFilename)) {
+      this.filename = newFilename;
+    }
+  }
+
+  /** @param secretJsonKey json key file content.*/
+  @DataBoundSetter
+  public void setSecretJsonKey(SecretBytes secretJsonKey) {
+    if (secretJsonKey != null && secretJsonKey.getPlainData().length > 0) {
+      this.secretJsonKey = secretJsonKey;
+    }
+  }
+
+
+  @Deprecated
+  // Used for JsonServiceAccountConfig
+  public JsonServiceAccountConfig(String prevJsonKeyFile) {
+    this.filename = extractFilename(prevJsonKeyFile);
+    this.secretJsonKey = getSecretBytesFromFile(prevJsonKeyFile);
+  }
```

Comment:
Rather you’d better add a constructor with the signature same to the existing one. It preserves binary compatibilities.: https://wiki.eclipse.org/Evolving_Java-based_APIs_2

Reply:
Makes sense, I'll revert changing these signatures and just add new mosty empty DataboundConstructors.